### PR TITLE
Fix a crash by setting not exist material parameters on ubershader

### LINKED ISSUE #6109: clearCoatMap : reason: sampler named "clearCoatMap" not found
**Describe the bug**
clearCoatMap : reason: sampler named "clearCoatMap" not found

**To Reproduce**
1. Import a new glb model into the project hello-gltf in Maos 12 to replace the original model

**Expected behavior**
crash : reason: sampler named "clearCoatMap" not found

**Screenshots**
Screen Shot 2022-09-28 at 11 19 49

**Logs**
If applicable, copy logs from your console here. Please *do not*
use screenshots of logs, copy them as text.
2022-09-28 11:18:30.982812+0800 hello-gltf[4591:105410] Metal GPU Frame Capture Enabled
2022-09-28 11:18:30.983199+0800 hello-gltf[4591:105410] Metal API Validation Enabled
FEngine (64 bits) created at 0x7fe5b180c400 (threading is enabled)
FEngine resolved backend: Metal
Selected physical device 'Apple iOS simulator GPU'
FEngine feature level: 1
Panic
in getSamplerInfo:97
reason: sampler named "clearCoatMap" not found
CoreSimulator 842.8 - Device: iPhone 13 (D1A3AAD9-E8DE-4E05-9889-8D12FB891708) - Runtime: iOS 16.0 (20A5303f) - DeviceType: iPhone 13

**Desktop (please complete the following information):**
 - OS: [iOS]
 - GPU: [AMD]
 - Backend: [Metal]

**Smartphone (please complete the following information):**
 - Device: [MacOS12/ iPhone]
 - OS: [iOS16]
 - Xcode: [14]
 - filament: [filament-1.27.1]

**Additional context**
I found that this problem is very similar, but the original library file seems to be missing. I don't know how to deal with it in the new version
[https://github.com/google/filament/issues/4143](https://github.com/google/filament/issues/4143)

### LINKED ISSUE #6121: Crash an application when setting not exist material parameters on ubershader
**Describe the bug**
Crash an application when setting not exist material parameters. Material comes from ubershader.

**To Reproduce**
1. Prepare a GLTF model that has material properties. This material causes a crash.
2. Run gltf_viewer with ubershader

Sorry I can't share this model file. But this crash is happened if `hasIOR` is true.
```
(filament::gltfio::MaterialKey) $0 = {
  doubleSided = true
  unlit = false
  hasVertexColors = true
  hasBaseColorTexture = false
  hasNormalTexture = false
  hasOcclusionTexture = false
  hasEmissiveTexture = false
  useSpecularGlossiness = false
  alphaMode = OPAQUE
  enableDiagnostics = false
   = {
     = (hasMetallicRoughnessTexture = false, metallicRoughnessUV = 0 '\0')
     = (hasSpecularGlossinessTexture = false, specularGlossinessUV = 0 '\0')
  }
  baseColorUV = '\0' 0 '\0'
  hasClearCoatTexture = false
  clearCoatUV = '\0' 0 '\0'
  hasClearCoatRoughnessTexture = false
  clearCoatRoughnessUV = '\0' 0 '\0'
  hasClearCoatNormalTexture = false
  clearCoatNormalUV = '\0' 0 '\0'
  hasClearCoat = false
  hasTransmission = false
  hasTextureTransforms = false
  emissiveUV = '\0' 0 '\0'
  aoUV = '\0' 0 '\0'
  normalUV = '\0' 0 '\0'
  hasTransmissionTexture = false
  transmissionUV = '\0' 0 '\0'
  hasSheenColorTexture = false
  sheenColorUV = '\0' 0 '\0'
  hasSheenRoughnessTexture = false
  sheenRoughnessUV = '\0' 0 '\0'
  hasVolumeThicknessTexture = false
  volumeThicknessUV = '\0' 0 '\0'
  hasSheen = false
  hasIOR = true
  hasVolume = false
  padding = '\0' 0 '\0'
}
```

**Expected behavior**
Load a GLTF model and run well.

**Logs**
```
utils::PreconditionPanic
in const SamplerInterfaceBlock::SamplerInfo *filament::SamplerInterfaceBlock::getSamplerInfo(std::string_view) const:97
in file filament/libs/filabridge/src/SamplerInterfaceBlock.cpp
reason: sampler named "clearCoatMap" not found
#0	0x1011147ac utils::CallStack::update(unsigned long) + 32
#1	0x1011308d0 utils::TPanic<utils::PreconditionPanic>::TPanic(char const*, char const*, int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) + 168
#2	0x101131c84 utils::PreconditionPanic::PreconditionPanic(char const*, char const*, int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) + 68
#3	0x101130478 utils::PreconditionPanic::PreconditionPanic(char const*, char const*, int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) + 52
#4	0x101130244 utils::TPanic<utils::PreconditionPanic>::panic(char const*, char const*, int, char const*, ...) + 228
#5	0x101109714 filament::SamplerInterfaceBlock::getSamplerInfo(std::__1::basic_string_view<char, std::__1::char_traits<char> >) const + 176
#6	0x100fa55ec filament::FMaterialInstance::setParameter(std::__1::basic_string_view<char, std::__1::char_traits<char> >, filament::backend::Handle<filament::backend::HwTexture>, filament::backend::SamplerParams) + 76
#7	0x100fa57e0 filament::FMaterialInstance::setParameterImpl(std::__1::basic_string_view<char, std::__1::char_traits<char> >, filament::FTexture const*, filament::TextureSampler const&) + 428
#8	0x100f024c8 filament::MaterialInstance::setParameter(char const*, unsigned long, filament::Texture const*, filament::TextureSampler const&) + 92
#9	0x100fcd6bc filament::MaterialInstance::setParameter(char const*, filament::Texture const*, filament::TextureSampler const&) + 76
#100x101241988 (anonymous namespace)::UbershaderProvider::createMaterialInstance(filament::gltfio::MaterialKey*, std::__1::array<filament::gltfio::UvSet, 8ul>*, char const*, char const*) + 2080
#110x10118c858 filament::gltfio::FAssetLoader::createMaterialInstance(cgltf_data const*, cgltf_material const*, std::__1::array<filament::gltfio::UvSet, 8ul>*, bool) + 452
#120x101189dd4 filament::gltfio::FAssetLoader::createRenderable(cgltf_data const*, cgltf_node const*, utils::Entity, char const*) + 656
#130x101189160 filament::gltfio::FAssetLoader::recurseEntities(cgltf_data const*, cgltf_node const*, utils::bitset<unsigned int, 1ul, void>, utils::Entity, filament::gltfio::FFilamentInstance*) + 964
#140x101189278 filament::gltfio::FAssetLoader::recurseEntities(cgltf_data const*, cgltf_node const*, utils::bitset<unsigned int, 1ul, void>, utils::Entity, filament::gltfio::FFilamentInstance*) + 1244
#150x101187b9c filament::gltfio::FAssetLoader::createInstance(cgltf_data const*) + 528
#160x101187560 filament::gltfio::FAssetLoader::createInstances(cgltf_data const*, unsigned long) + 72
#170x101186c2c filament::gltfio::FAssetLoader::createInstancedAsset(unsigned char const*, unsigned int, filament::gltfio::FilamentInstance**, unsigned long) + 416
```

**Additional context**
If the material key is like below
```
(filament::gltfio::MaterialKey) $0 = {
  doubleSided = true
  unlit = false
  hasVertexColors = true
  hasBaseColorTexture = false
  hasNormalTexture = false
  hasOcclusionTexture = false
  hasEmissiveTexture = false
  useSpecularGlossiness = false
  alphaMode = OPAQUE
  enableDiagnostics = false
   = {
     = (hasMetallicRoughnessTexture = false, metallicRoughnessUV = 0 '\0')
     = (hasSpecularGlossinessTexture = false, specularGlossinessUV = 0 '\0')
  }
  baseColorUV = '\0' 0 '\0'
  hasClearCoatTexture = false
  clearCoatUV = '\0' 0 '\0'
  hasClearCoatRoughnessTexture = false
  clearCoatRoughnessUV = '\0' 0 '\0'
  hasClearCoatNormalTexture = false
  clearCoatNormalUV = '\0' 0 '\0'
  hasClearCoat = false
  hasTransmission = false
  hasTextureTransforms = false
  emissiveUV = '\0' 0 '\0'
  aoUV = '\0' 0 '\0'
  normalUV = '\0' 0 '\0'
  hasTransmissionTexture = false
  transmissionUV = '\0' 0 '\0'
  hasSheenColorTexture = false
  sheenColorUV = '\0' 0 '\0'
  hasSheenRoughnessTexture = false
  sheenRoughnessUV = '\0' 0 '\0'
  hasVolumeThicknessTexture = false
  volumeThicknessUV = '\0' 0 '\0'
  hasSheen = false
  hasIOR = true
  hasVolume = false
  padding = '\0' 0 '\0'
}
```
then a queried material from ubershader has the sampler interface block like below.
```
(const filament::SamplerInterfaceBlock) $0 = {
  mName = {
     = {
      mCStr = 0x00006000009a0e44 "MaterialParams"
      mData = 0x00006000009a0e44
    }
  }
  mStageFlags = ALL_SHADER_STAGE_FLAGS
  mSamplersInfoList = {
    mData = 0x0000600003efb860
    mSize = 7
    mCapacityAllocator = {
      utils::FixedCapacityVector<filament::SamplerInterfaceBlock::SamplerInfo, std::__1::allocator<filament::SamplerInterfaceBlock::SamplerInfo>, true>::SizeTypeWrapper<unsigned int> = (value = 7)
    }
  }
  mInfoMap = size=7 {
    [0] = (first = "volumeThicknessMap", second = 6)
    [1] = (first = "transmissionMap", second = 5)
    [2] = (first = "normalMap", second = 2)
    [3] = (first = "occlusionMap", second = 3)
    [4] = (first = "metallicRoughnessMap", second = 1)
    [5] = (first = "emissiveMap", second = 4)
    [6] = (first = "baseColorMap", second = 0)
  }
}
```

### PR DESCRIPTION
**Pull request by daemyung**
Fix #6121 
Fix #6109

### CODE REVIEW

**Comment by romainguy** [comment]
That's not enough. You also need to fix it in `UberShaderProvider.cpp` by changing this line:

```
const bool clearCoatConflict = config->hasVolume || config->hasTransmission || config->hasSheen;
```

And the lines below where hasVolume, etc. are set to false.

**Comment by daemyung** [comment]
@romainguy IOR can be used with the clear coat isn't it? I read [Lighting: clearCoatIorChange](https://google.github.io/filament/Materials.html#materialdefinitions/materialblock/lighting:clearcoatiorchange).

**Comment by romainguy** [comment]
Currently not with the glTF Ubershader because of sampler count limits on mobile.

With custom materials there's no issue.

Note that the behavior you linked to can be controlled with the reflectance attribute as well.

**Comment by daemyung** [comment]
@romainguy Ok I understand. Changes should be like this right?

```
const bool clearCoatConflict = config->hasVolume || config->hasTransmission || config->hasSheen || config->hasIOR;

// Due to sampler overload, disable transmission if necessary and print a friendly warning.
if (config->hasClearCoat && clearCoatConflict) {
    slog.w << "Volume, transmission, sheen and IOR are not supported in ubershader mode for clearcoat"
              " materials (" << label << ")." << io::endl;
    config->hasVolume = false;
    config->hasTransmission = false;
    config->hasSheen = false;
    config->hasIOR = false;
}
`
``

But I still don't fully understand why more fixes are needed. Because IOR looks like an uniform variable and doesn't count the number of sampler.

**Comment by romainguy** [comment]
Yeah in theory we don't need to do this but any material property written to by the ubershader implies extra computations. Since IOR is not a common one it's not worth supporting in the common path (clear coat/basic features).

**Comment by daemyung** [comment]
Ok I got it. It's updated.

**Comment by romainguy** [comment]
Thank you! I'll let @prideout do the final review. Maybe he knows of a way to add IOR to the common path at no cost.

**Comment by prideout** [comment]
Actually, Romain wrote an ubershader that can handle this model just fine (it's called `volume.mat.in`).

I took a look and the ArchiveCache is properly choosing this ubershader, but `UbershaderProvider::createMaterialInstance` has faulty logic that causes it to try to set the CC texture when in fact there is no CC texture.

I think a better fix is to simply change this:

```
bool clearCoatNeedsTexture = true;
```

to

```
bool clearCoatNeedsTexture = config->hasClearCoatTexture;
```

**Comment by romainguy** [comment]
Yeah I did write a material for IOR but it doesn't handle clear coat *and* IOR together. This means we still need to this properly:

`const bool clearCoatConflict = config->hasVolume || config->hasTransmission || config->hasSheen || config->hasIOR;`

**Comment by prideout** [comment]
Yes, we still need that particular codeline for graceful degradation.  Eventually we should add an uberz spec property for FEATURE_LEVEL_3, for these devices we would bypass `prepareConfig()` and add a truly "uber" ubershader to the uberz archive.

The logic in `createMaterialInstance` is hard to follow because it's trying to deduce if the material has certain texture parameters.  I think we can get rid of that logic if we add a query to `ArchiveCache`.  Let me give that a shot.

**Comment by prideout** [comment]
@daemyung @romainguy I've added a commit to this PR that replaces the guesswork for dummy textures.  Romain's suggest fix is still there too.  I think this is in better shape now.

**Comment by daemyung** [comment]
@prideout Your shot works well. :)

**Review by prideout — APPROVED** [review]

**Comment by romainguy** [comment]
Thank you @daemyung!
